From: Ahmad Fatoum <a.fatoum@pengutronix.de>
Date: Tue, 14 May 2024 06:49:29 +0200
Subject: [PATCH] ARM: i.MX8MP: skov: halt startup until power is good

The 24V regulator supplying the system can indicate via a GPIO imminent
voltage loss. The board's capacitors hold enough charge to power the system
a while longer in such a state, but eventually, unless external power is
restored, the brownout detection of the PMIC will kick in.

For the span of time between voltage drop detection and PMIC brownout,
let's detect this situation and delay startup. This way, Linux can
detect the ongoing voltage loss, power down the eMMC, reboot into
barebox and barebox will delay boot as long as the problem persists.

Signed-off-by: Ahmad Fatoum <a.fatoum@pengutronix.de>
---
 arch/arm/boards/skov-imx8mp/lowlevel.c | 34 +++++++++++++---------------------
 1 file changed, 13 insertions(+), 21 deletions(-)

diff --git a/arch/arm/boards/skov-imx8mp/lowlevel.c b/arch/arm/boards/skov-imx8mp/lowlevel.c
index 1fd665d67675..6a4cb72ae38e 100644
--- a/arch/arm/boards/skov-imx8mp/lowlevel.c
+++ b/arch/arm/boards/skov-imx8mp/lowlevel.c
@@ -85,18 +85,15 @@ static inline bool power_good(void)
 static void wait_for_power_good(void)
 {
 	void __iomem *gpio4 = IOMEM(MX8MP_GPIO4_BASE_ADDR);
-	int timeout_ms = 0;
-	bool led_active = true;
+	int timeout_ms = 100;
 
 	imx8mp_setup_pad(MX8MP_PAD_SAI2_RXD0__GPIO4_IO23 | PGOOD_PAD_CTRL);
 	imx8m_gpio_direction_input(gpio4, 23);
 
-	led_d1_toggle(&led_active);
-
 	if (power_good())
 		return;
 
-	pr_warn("\nDelaying boot until power stabilizes\n");
+	pr_debug("Delaying boot until power stabilizes");
 
 	/* If we reach this, because Linux did a hw_protection_reboot, we don't
 	 * want to continue booting right away.
@@ -105,24 +102,19 @@ static void wait_for_power_good(void)
 	 * to reach a low enough level for the PMIC to detect VSYS_UVLO going
 	 * lower than allowed
 	 */
-
-	while (1) {
-		if (power_good()) {
-			/* wait 10ms longer and check if it still good */
-			udelay(10000);
-			if (power_good()) {
-				pr_info("IMX_SHDN_MF stuck low for ~%ums.\n", timeout_ms);
-				break;
-			}
-		}
-		/* fast blink LED D1 */
-		if (timeout_ms % 100 == 0) {
-			pr_debug(".");
-			led_d1_toggle(&led_active);
-		}
+	while (!power_good() && timeout_ms > 0) {
+		pr_debug(".");
 		udelay(1000);
-		timeout_ms++;
+		timeout_ms--;
 	}
+
+	if (!timeout_ms)
+		pr_warn("\nIMX_SHDN_MF stuck low for >%ums. Continuing anyway\n",
+			timeout_ms);
+	/* The else branch is never reached, because power good toggling also
+	 * resets the SoC
+	 */
+
 }
 
 static void power_init_board(void)
